GCE API Tests - Part 1

jyejare opened the pull request:
$subject

mirekdlugosz commented on tests/foreman/api/test_computeresource_gce.py:
```diff
@@ -0,0 +1,250 @@
+        assert GCEnvironment.finishimg(
+            gceclient).compute_resource.id == GCEnvironment.gce_compute().id
+        assert GCEnvironment.finishimg(gceclient).uuid == GCEnvironment.latest_rhel_uuid()
+        assert GCEnvironment.finishimg().username == GCEnvironment.finishimg_user
```
This is one of weaknesses of that wrapper and API - it's not clear where test data comes from and what exactly is verified here. Since `GCEnvironment` controls both `finishimg` and `finishimg_user`, it looks more like unit test for wrapper that verifies if certain data is available through two different interfaces.

jyejare replied:
That can be fixed by passing a username parameter. But That way I will end up with multiple usernames and this being a helper function which sets only once, I dont want to do that.

mirekdlugosz commented on tests/foreman/api/test_computeresource_gce.py:
```diff
@@ -0,0 +1,250 @@
+
+    @tier1
+    def test_positive_host_provisioned(self):
+        assert self.host
```
Can this assertion ever be false? `host` is created in class setup. If `Host.create()` throws exception, that test will not even run. And as far as I know, there is no way for `Host.create()` to silently return falsey value.

jyejare replied:
Removed.

mirekdlugosz commented on tests/foreman/api/test_computeresource_gce.py:
```diff
@@ -0,0 +1,250 @@
+            operatingsystem=GCEnvironment.operatingsystem(), location=GCEnvironment.location(),
+            name=cls.hostname, provision_method='image', root_pass=gen_string('alphanumeric')
+        ).create()
+        cls.fullhostname = '{}.{}'.format(cls.hostname, GCEnvironment.domain().name).lower()
```
That should be specified in relevant tests, not in class setup. Setup is reserved for time-consuming operations where results can be used by multiple tests.

jyejare replied:
Probably not in current tests, but it will be used multiple times in future.

mirekdlugosz commented on tests/foreman/api/test_computeresource_gce.py:
```diff
@@ -0,0 +1,250 @@
+
+    @tier3
+    def test_positive_power_on_off(self):
+        self.host.power()
```
Shouldn't you add `power_action="off"` argument? At least on Nailgun side, there doesn't seem to be any magic to decide which action should be called.

jyejare replied:
Half baked. Now the correct is there.

jyejare commented:
@mirekdlugosz , Thanks for all the comments, and extremely sorry for not putting Do_NOT_REVIEW or MERGE on PR. The PR is being worked on. I will have latest updates for you soon.

jyejare commented:
Test Results:
```
================ 1 failed, 6 passed, 2 skipped in 482.46 seconds =========
```

jyejare commented on tests/foreman/api/test_computeresource_gce.py:
```diff
@@ -0,0 +1,250 @@
+
+        :id:
+
+        :expectedresults: GCE Compute resource should be created
```
updated!

jyejare commented on tests/foreman/api/test_computeresource_gce.py:
```diff
@@ -0,0 +1,250 @@
+            GCEnvironment.gce_compute().available_networks()
+        ) == len(
+            self.gclient.list_network()
+        )
```
I prefer to not set a variable which won't be used multiple lines. still, I updated them for more clearity of assertions.

jyejare commented on tests/foreman/api/test_computeresource_gce.py:
```diff
@@ -0,0 +1,250 @@
+        assert len(
+            GCEnvironment.gce_compute().available_networks()
+        ) == len(
+            self.gclient.list_network()
```
The half baked PR comment. Now it has a correct thing.

jyejare commented on tests/foreman/api/test_computeresource_gce.py:
```diff
@@ -0,0 +1,250 @@
+
+        :CaseLevel: Integration
+        """
+        assert len(GCEnvironment.gce_compute().available_flavors()) > 1
```
There is Bug [1] Raised in BZ to get the exact count of flavors. Currently the count from Satellite, Google Cloud UI and WrappenAPI Google Cloud API, all of them show different count.

So think it as just a placeholder and the Test is already being skipped with that BZ now.

jyejare commented:
> All tests miss Id metadata, which cause Travis to fail. Of course these must be added.

Half baked PR. Now all has Id's.
> Overall this looks like heading in right direction, despite number of comments that I added. It adds coverage, which is obviously a good thing.

Great to hear that.
> I am not sold on `robottelo/api/provisioning.py` idea and implementation.
> 
>     * It duplicates some of the wrappers that we already have. I would rather re-use them, so if there is need for changes in future, we can change only one wrapper.

Are you talking about :

- robottelo/api/utils -> configure_provisioning: This sets up katello components as well which is not required for image-based provisioning. Hence It was required to separate Global Entities, Katello Entities and ComputeResource specific Entities, so that all of them can be used as per once convenience.
> 
>     * It is very heavy on singletons and deferred execution. Given these objects are supposed to be global in context of couple of test classes, it seems unnecessary complication. Just create all entities on class initialization.

Yeah, the scope part can be improved by ```object.methods``` instead of ```class.methods``` where we can initiate an object classwise. Should I raise a different PR for it or are you insisting on doing in this PR only ?
> 
>     * I believe these wrappers try to do too much, to the point that it's unclear what is the point of some of the tests. I left comment in on of them.

I fixed most of them. Probably it was there because the PR was half baked. Let me know if you have more.
>
>     * In general, that wrapper should be implemented in class setup or module-scoped fixture.

Need more context or what wrapper your specific on ?

jyejare commented:
@mirekdlugosz Hold the review as we are updating the provisioning library as fixtures.

jyejare commented:
@mirekdlugosz , @vijay8451 :

Results post fixturization:
```
$ py.test tests/foreman/api/test_computeresource_gce.py
========================================== test session starts ==========================================
platform linux -- Python 3.7.4, pytest-4.6.3, py-1.8.0, pluggy-0.12.0
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/jitendrayejare/Desktop/RedHat/RoboTelloNew/robottelo
plugins: forked-1.0.2, xdist-1.29.0, cov-2.7.1, services-1.3.1, mock-1.10.4
collecting ... 2020-01-30 10:59:23 - conftest - DEBUG - Collected 9 test cases
2020-01-30 16:29:23 - robottelo.issue_handlers.bugzilla - DEBUG - Calling Bugzilla API for {'1794744'}
2020-01-30 16:29:34 - robottelo.ssh - DEBUG - Instantiated Paramiko client 0x7f23008da350
2020-01-30 16:29:34 - robottelo.ssh - INFO - Connected to [qe-sat67-rhel7-tier3.satqe.lab.eng.rdu2.redhat.com]
2020-01-30 16:29:34 - robottelo.ssh - INFO - >>> rpm -q satellite
2020-01-30 16:29:36 - robottelo.ssh - INFO - <<< stdout
satellite-6.7.0-5.beta.el7sat.noarch

2020-01-30 16:29:36 - robottelo.ssh - DEBUG - Destroyed Paramiko client 0x7f23008da350
2020-01-30 16:29:36 - robottelo.host_info - DEBUG - Host Satellite version: 6.7
2020-01-30 16:29:36 - robottelo.helpers - DEBUG - Generated file bz_cache.json with BZ collect data
collected 9 items                                                                                       

tests/foreman/api/test_computeresource_gce.py .s.s.....                                           [100%]

================================= 7 passed, 2 skipped in 444.26 seconds =================================
```

Also, PR is ready for review. and thats final I promise. :)

mirekdlugosz commented on conftest.py:
```diff
@@ -1,2 +1,201 @@
+def gce_latest_rhel_uuid(googleclient):
+    max_rhel7_template = max(
+        [img.name for img in googleclient.list_templates(
+            True) if str(img.name).startswith('rhel-7')])
```
List comprehension doesn't really serve readability of this. I would break it into two lines:
```python
template_names = [img.name for img in googleclient.list_templates(True)]
latest_rhel7_template = max(name for name in template_names if name.startswith('rhel-7'))
```

jyejare replied:
Again, I dont waste variable if its not reusable, but updating for better understanding.

mirekdlugosz commented on robottelo/helpers.py:
```diff
@@ -1080,8 +1080,10 @@ def _handle_version(key, value):  # pragma: no cover
-    if ssh.command('[ -f {} ]'.format(gce_cert_path)).return_code != 0:
+def download_gce_cert():
+    ssh.command('curl {0} -o {1}'.format(settings.gce.cert_url, settings.gce.cert_path))
+    if not ssh.command('[ -f {} ]'.format(settings.gce.cert_path)).return_code == 0:
```
In this line, previous way was more natural. It was "if command return is not equal zero" instead of "if not command return equals zero".

mirekdlugosz commented on tests/foreman/api/test_computeresource_gce.py:
```diff
@@ -0,0 +1,357 @@
+
+        :id: 259c9f5f-19dd-4e06-af5e-874612914f44
+
+        :expectedresults: All the networks from Google CR should be available to select in GCE CR
```
Technically, this only tests if number of networks is the same, not if they are actually the same networks.

Does `module_gce_compute.available_networks()` provide list of networks as well? Perhaps you could put that into `set()` and compare against `set()` of `googleclient.list_network()`?

mirekdlugosz commented on tests/foreman/api/test_computeresource_gce.py:
```diff
@@ -0,0 +1,357 @@
+
+        :id: 5cdfab18-a591-4442-8c19-a01e9b10ac36
+
+        :expectedresults: All the images from Google CR should be available to select in GCE CR
```
Same here - technically we only check if number of images is the same, not if these are the same images.

jyejare replied:
Actually what Google Compute returns is different in names is different than what Foreman GCE shows, Slight change in name. But I will be able to confirm/ fix this post the BZ bug for the test is fixed. This is just a placeholder for now.

mirekdlugosz commented on tests/foreman/api/test_computeresource_gce.py:
```diff
@@ -0,0 +1,357 @@
+    :CaseComponent: Hosts
+    """
+    @pytest.fixture(scope='class')
+    def gce_domain(self, module_org, module_location, module_smart_proxy):
```
Can you move these fixtures to top of file? Usually fixtures are defined as global functions, not as methods of class. having them here is unexpected.

jyejare replied:
Not mandatory to Keep in Module Scope but yes the newly added tests (by new contributor) for that person/tests it will be helpful.

Moving at module level.

jyejare commented on conftest.py:
```diff
@@ -1,2 +1,201 @@
 # coding: utf-8
```
Yes, So I wanted to put that in tests/foreman, so that it will be used API,CLI and UI tests. Thanks for confirmation.

jyejare commented:
Post Changes result:
```
===================== 7 passed, 2 skipped in 367.76 seconds ==================
```

vijay8451 commented on tests/foreman/api/test_computeresource_gce.py:
```diff
@@ -0,0 +1,360 @@
+            architecture=module_architecture, compute_attributes=self.compute_attrs,
+            domain=gce_domain, hostgroup=gce_hostgroup, organization=module_org,
+            operatingsystem=module_os, location=module_location, name=self.hostname,
+            provision_method='image', image=module_gce_finishimg,
```
Do you plan to cover `user data/cloud-init ` host create as well ?

jyejare replied:
The cloud-init has been covered by UI plus the CLI automation would be based on Cloud-init.
